### ISSUE TITLE
pyUSB installed? 

### ISSUE THREAD

**Issue by Nornode** [felddy/weewx-docker#16]
A well documented and pretty straight forward container, finally!
Thanks for your job, but which weather station works for you and how is it connected?

I'm running this on an Rpi2 but get an error:
Possibly for a needed dependency for a USB connection.

```
May 26 12:20:12 syslogd started: BusyBox v1.31.1
May 26 12:20:15 weewx[12] INFO __main__: Initializing weewx version 4.0.0
May 26 12:20:15 weewx[12] INFO __main__: Using Python 3.8.3 (default, May 20 2020, 19:12:37)  [GCC 9.2.0]
May 26 12:20:15 weewx[12] INFO __main__: Platform Linux-4.19.106-armv7l-with
May 26 12:20:15 weewx[12] INFO __main__: Locale is 'C.UTF-8;C;C;C;C;C'
May 26 12:20:15 weewx[12] INFO __main__: Using configuration file /data/weewx.conf
May 26 12:20:15 weewx[12] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
May 26 12:20:15 weewx[12] CRITICAL __main__: Caught unrecoverable exception:
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****  No module named 'usb'
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****  Traceback (most recent call last):
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****    File "./bin/weewxd", line 148, in main
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****      engine = weewx.engine.StdEngine(config_dict)
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 69, in __init__
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****      self.setupStation(config_dict)
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 92, in setupStation
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****      __import__(driver)
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 219, in <module>
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****      import usb
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****  ModuleNotFoundError: No module named 'usb'
May 26 12:20:15 weewx[12] CRITICAL __main__:     ****  Exiting.
Traceback (most recent call last):
  File "./bin/weewxd", line 261, in <module>
    main()
  File "./bin/weewxd", line 148, in main
    engine = weewx.engine.StdEngine(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 69, in __init__
    self.setupStation(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 92, in setupStation
    __import__(driver)
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 219, in <module>
    import usb
ModuleNotFoundError: No module named 'usb'
```
I'm starting my container this way:
```
docker create \
  -i \
  -t \
  --name=$name \
  --hostname=$name \
  --net=host \
  --privileged \
  --stop-signal=SIGRTMIN+3 \
  --tmpfs /run:size=100M \
  --tmpfs /run/lock:size=10M \
  -v /storage/nfs.docker/$name:/data \
  --restart unless-stopped \
  -e TIMEZONE=Europe/Helsinki \
  -e WEEWX_UID=weewx \
  -e WEEWX_GID=dialout \
    felddy/weewx
```

Is pyUSB installed and can be used? 
My weatherization is a FineOffset WH1080 and Docker version 18.09.7, build 18.09.7

**Comment by felddy** [comment]
Thanks for your issue submission.  

`pyusb` was not part of the container.  I'm testing the addition now.
I'm using a `Davis Vantage Vue (Wireless)` over a USB serial connection.  So it appears as a serial port on my Linux machine which I map into the container as a `device`.  I expect you'll need to do something similar with your USB station.  

For reference I run it using `docker-compose` and this `docker-compose.yml` file:

```yaml
---
version: "3.7"

volumes:
  data:

services:
  weewx:
    image: felddy/weewx
    init: true
    restart: always
    volumes:
      - type: bind
        source: ./data
        target: /data
    environment:
      - TIMEZONE=US/Eastern
      - WEEWX_UID=weewx
      - WEEWX_GID=dialout
    devices:
      - /dev/ttyUSB0
```

Once testing is done, I'll build a new pre-release and comment here.  I won't be able to test it, so I would appreciate any feedback you have.  

Stay tuned.

**Comment by felddy** [comment]
@Nornode , I've pre-released a new image out to Docker hub.  If you pull `latest` you should get it. 

https://github.com/felddy/weewx-docker/releases/tag/v4.0.0-pre.2%2Busb

Please test and let me know if it solves your issue.

**Comment by Nornode** [comment]
Awesome, thanks for super quick response!
I managed to pull the image but some error still remains.

```
May 26 22:17:09 syslogd started: BusyBox v1.31.1
May 26 22:17:12 weewx[13] INFO __main__: Initializing weewx version 4.0.0
May 26 22:17:12 weewx[13] INFO __main__: Using Python 3.8.3 (default, May 20 2020, 19:12:37)  [GCC 9.2.0]
May 26 22:17:12 weewx[13] INFO __main__: Platform Linux-4.19.106-armv7l-with
May 26 22:17:12 weewx[13] INFO __main__: Locale is 'C.UTF-8;C;C;C;C;C'
May 26 22:17:12 weewx[13] INFO __main__: Using configuration file /data/weewx.conf
May 26 22:17:12 weewx[13] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
May 26 22:17:12 weewx[13] INFO weewx.drivers.fousb: driver version is 1.20
May 26 22:17:12 weewx[13] INFO weewx.drivers.fousb: polling mode is PERIODIC
May 26 22:17:12 weewx[13] INFO weewx.drivers.fousb: polling interval is 60
May 26 22:17:13 weewx[13] ERROR weewx.engine: Import of driver failed: No backend available (<class 'usb.core.NoBackendError'>)
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****  Traceback (most recent call last):
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/engine.py", line 103, in setupStation
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      self.console = loader_function(config_dict, self)
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 233, in loader
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      return FineOffsetUSB(**config_dict[DRIVER_NAME])
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 968, in __init__
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      self.openPort()
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 1019, in openPort
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      dev = self._find_device()
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 1052, in _find_device
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      for bus in usb.busses():
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/legacy.py", line 353, in busses
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      sorted(core.find(find_all=True), key=lambda d: d.bus),
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/core.py", line 1263, in find
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****      raise NoBackendError('No backend available')
May 26 22:17:13 weewx[13] CRITICAL weewx.engine:     ****  usb.core.NoBackendError: No backend available
May 26 22:17:13 weewx[13] CRITICAL __main__: Unable to load driver: No backend available
May 26 22:17:13 weewx[13] CRITICAL __main__:     ****  Exiting...
```

According to this post: https://groups.google.com/forum/#!topic/weewx-user/sY5TJijPPv8
Is the container still missing "the underlying usb "C" library" which translates into _python-usb_?

I wished that I could use docker-compose to make the initiation as much the same as possible, unfortunately it's not available on the OS I run this.

**Comment by felddy** [comment]
@Nornode With regards to `docker-compose`... you can install it with `pip`:  

```console
pip install docker-compose
```

I've created a new release and pushed it up.   I had to make some substantial changes.  The previous images were based on Alpine Linux.  It wasn't playing well with `pyusb` and `libusb` so I've switched to a Debian base.  

https://github.com/felddy/weewx-docker/releases/tag/v4.0.0-pre.2%2Busb

Please pull the new image and try it out.  I was able to get Python to successfully load the backend using the test code here:

https://github.com/walac/pyusb/blob/master/docs/faq.rst

Let me know if it works for you.

**Comment by Nornode** [comment]
Wow, some impressive changes you made! 
Much, much appreciated!

Unfortunately I think I'm back with a regressed error:

I pulled: 
```
docker pull felddy/weewx:latest
latest: Pulling from felddy/weewx
Digest: sha256:b7a72da239211c5869a4a6dc72694ed8854ee34b77c4b3fff3e2519e05885913
Status: Image is up to date for felddy/weewx:latest
```
Start log output from container...
```
May 27 09:15:20 syslogd started: BusyBox v1.31.1
May 27 09:15:24 weewx[13] INFO __main__: Initializing weewx version 4.0.0
May 27 09:15:24 weewx[13] INFO __main__: Using Python 3.8.3 (default, May 20 2020, 19:12:37)  [GCC 9.2.0]
May 27 09:15:24 weewx[13] INFO __main__: Platform Linux-4.19.106-armv7l-with
May 27 09:15:24 weewx[13] INFO __main__: Locale is 'C.UTF-8;C;C;C;C;C'
May 27 09:15:24 weewx[13] INFO __main__: Using configuration file /data/weewx.conf
May 27 09:15:24 weewx[13] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
May 27 09:15:24 weewx[13] CRITICAL __main__: Caught unrecoverable exception:
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****  No module named 'usb'
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****  Traceback (most recent call last):
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****    File "./bin/weewxd", line 148, in main
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****      engine = weewx.engine.StdEngine(config_dict)
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 69, in __init__
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****      self.setupStation(config_dict)
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 92, in setupStation
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****      __import__(driver)
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 219, in <module>
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****      import usb
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****  ModuleNotFoundError: No module named 'usb'
May 27 09:15:24 weewx[13] CRITICAL __main__:     ****  Exiting.
Traceback (most recent call last):
  File "./bin/weewxd", line 261, in <module>
    main()
  File "./bin/weewxd", line 148, in main
    engine = weewx.engine.StdEngine(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 69, in __init__
    self.setupStation(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 92, in setupStation
    __import__(driver)
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 219, in <module>
    import usb
ModuleNotFoundError: No module named 'usb'
```

**Comment by felddy** [comment]
@Nornode Well that is strange... could you please post the output of this command:

```console
docker inspect -f {{.Config.Labels}} felddy/weewx
```

**Comment by Nornode** [comment]
```javascript
docker inspect -f {{.Config.Labels}} felddy/weewx
map[org.opencontainers.image.authors:markf+github@geekpad.com 
org.opencontainers.image.licenses:CC0-1.0 
org.opencontainers.image.revision:8fb01a2a4e9ead1e5ca8e6652044cac1782d5e04 
org.opencontainers.image.source:https://github.com/felddy/weewx-docker 
org.opencontainers.image.title:WeeWX
org.opencontainers.image.vendor:Geekpad 
org.opencontainers.image.version:4.0.0]
```

**Comment by felddy** [comment]
@Nornode Thanks for the output. 
That image's revision is from the wrong branch: 8fb01a2a4e9ead1e5ca8e6652044cac1782d5e04
 Looks like I built `v4.0.0-pre.3+usb` off the `develop` branch by accident.  Sorry about that.  I just published `v4.0.0-pre.4+usb` from the correct branch.  

This new test image should be labeled with revision c79bf45d236ac5f0f8e838d06697ae8ef25d55d7. 

Please `pull` and test again.

**Comment by Nornode** [comment]
Okay, pulled and I think it's very close to be working! 

Yet, there are still some quirks...
Starting the container gives me the following:

```
Start log output from container...
May 27 21:25:14 syslogd started: BusyBox v1.30.1
May 27 21:25:18 weewx[19] INFO __main__: Initializing weewx version 4.0.0
May 27 21:25:18 weewx[19] INFO __main__: Using Python 3.8.3 (default, May 20 2020, 19:06:13)  [GCC 8.3.0]
May 27 21:25:18 weewx[19] INFO __main__: Platform Linux-4.19.106-armv7l-with-glibc2.4
May 27 21:25:18 weewx[19] INFO __main__: Locale is 'C.UTF-8'
May 27 21:25:18 weewx[19] INFO __main__: Using configuration file /data/weewx.conf
May 27 21:25:18 weewx[19] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
May 27 21:25:18 weewx[19] INFO weewx.drivers.fousb: driver version is 1.20
May 27 21:25:18 weewx[19] INFO weewx.drivers.fousb: polling mode is PERIODIC
May 27 21:25:18 weewx[19] INFO weewx.drivers.fousb: polling interval is 60
May 27 21:25:18 weewx[19] INFO weewx.drivers.fousb: found station on USB bus= device=
May 27 21:25:18 weewx[19] CRITICAL weewx.drivers.fousb: Unable to claim USB interface 0: [Errno 13] Access denied (insufficient permissions)
May 27 21:25:18 weewx[19] ERROR weewx.engine: Import of driver failed: [Errno 13] Access denied (insufficient permissions) (<class 'weewx.WeeWxIOError'>)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****  Traceback (most recent call last):
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 1037, in openPort
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      self.devh.claimInterface(self.usb_interface)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/legacy.py", line 230, in claimInterface
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      util.claim_interface(self.dev, interface)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/util.py", line 205, in claim_interface
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      device._ctx.managed_claim_interface(device, interface)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/core.py", line 102, in wrapper
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      return f(self, *args, **kwargs)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/core.py", line 159, in managed_claim_interface
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      self.managed_open()
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/core.py", line 102, in wrapper
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      return f(self, *args, **kwargs)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/core.py", line 120, in managed_open
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      self.handle = self.backend.open_device(self.dev)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/backend/libusb1.py", line 786, in open_device
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      return _DeviceHandle(dev)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/backend/libusb1.py", line 643, in __init__
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      _check(_lib.libusb_open(self.devid, byref(self.handle)))
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/opt/venv/lib/python3.8/site-packages/usb/backend/libusb1.py", line 595, in _check
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      raise USBError(_strerror(ret), ret, _libusb_errno[ret])
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****  usb.core.USBError: [Errno 13] Access denied (insufficient permissions)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****  During handling of the above exception, another exception occurred:
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****  Traceback (most recent call last):
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/engine.py", line 103, in setupStation
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      self.console = loader_function(config_dict, self)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 233, in loader
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      return FineOffsetUSB(**config_dict[DRIVER_NAME])
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 968, in __init__
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      self.openPort()
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****    File "/home/weewx/bin/weewx/drivers/fousb.py", line 1041, in openPort
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****      raise weewx.WeeWxIOError(e)
May 27 21:25:18 weewx[19] CRITICAL weewx.engine:     ****  weewx.WeeWxIOError: [Errno 13] Access denied (insufficient permissions)
May 27 21:25:18 weewx[19] CRITICAL __main__: Unable to load driver: [Errno 13] Access denied (insufficient permissions)
May 27 21:25:18 weewx[19] CRITICAL __main__:     ****  Exiting...
```
It leads me to think... Does the user running weewx process have root privileges?


Updated the start script to keep the container running:

```
docker create \
  -i \
  -t \
  --privileged \
  --name=$name \
  --hostname=$name \
  --net=host \
  --device=/dev/ttyUSB0 \
  --device=/dev/ttyAMA0 \
  -v /storage/nfs.docker/$name:/data \
  --restart unless-stopped \
  -e TIMEZONE=Europe/Helsinki \
  -e WEEWX_UID=weewx \
  -e WEEWX_GID=dialout \
  --entrypoint='tail' \
    felddy/weewx:latest -f /dev/null
```
Lo and behold! 
`root@weewx:/home/weewx/bin# ./wee_device /data/weewx.conf --info >> /data/output-wee_device2.txt`

```
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/logging/handlers.py", line 934, in emit
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/logging/handlers.py", line 855, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.8/logging/handlers.py", line 937, in emit
    self._connect_unixsocket(self.address)
  File "/usr/local/lib/python3.8/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "./wee_device", line 78, in <module>
    main()
  File "./wee_device", line 74, in main
    device.configure(config_dict)
  File "/home/weewx/bin/weewx/drivers/__init__.py", line 67, in configure
    self.do_options(options, parser, config_dict, not options.noprompt)
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 382, in do_options
    self.station = FineOffsetUSB(**config_dict[DRIVER_NAME])
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 960, in __init__
    log.info('driver version is %s' % DRIVER_VERSION)
Message: 'driver version is 1.20'
Arguments: ()
--- Logging error ---
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/logging/handlers.py", line 934, in emit
    self.socket.send(msg)
OSError: [Errno 9] Bad file descriptor

...
...
...
  File "/usr/local/lib/python3.8/logging/handlers.py", line 866, in _connect_unixsocket
    self.socket.connect(address)
FileNotFoundError: [Errno 2] No such file or directory
Call stack:
  File "./wee_device", line 78, in <module>
    main()
  File "./wee_device", line 74, in main
    device.configure(config_dict)
  File "/home/weewx/bin/weewx/drivers/__init__.py", line 67, in configure
    self.do_options(options, parser, config_dict, not options.noprompt)
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 382, in do_options
    self.station = FineOffsetUSB(**config_dict[DRIVER_NAME])
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 968, in __init__
    self.openPort()
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 1019, in openPort
    dev = self._find_device()
  File "/home/weewx/bin/weewx/drivers/fousb.py", line 1056, in _find_device
    log.info('found station on USB bus=%s device=%s' % (bus.dirname, dev.filename))
Message: 'found station on USB bus= device='
Arguments: ()
```
But Output into the file however looks "ok":
```
Using configuration file /data/weewx.conf
Using FineOffsetUSB driver version 1.20 (weewx.drivers.fousb)
Querying the station...
Fine Offset station settings:
                    local time: 2020.05.27 18:07:29 UTC
                  polling mode: PERIODIC

                  abs_pressure: 1021.3000000000001
                   current_pos: 352
                  data_changed: 0
                    data_count: 7
                     date_time: 2007-01-01 12:46
                 hum_in_offset: 17938
                hum_out_offset: 257
                            id: 46175
                 lux_wm2_coeff: 0
                       magic_1: 0x55
                       magic_2: 0xaa
                         model: 57656
                     rain_coef: 17152
                   read_period: 7
                  rel_pressure: 1013.0
                temp_in_offset: 1792
               temp_out_offset: 0
                      timezone: 32
                    unknown_01: 9
                    unknown_18: 0
                       version: 0
                     wind_coef: 39
                     wind_mult: 0

         max.abs_pressure.date: 2020-05-27 05:34
          max.abs_pressure.val: 1032.1000000000001
             max.dewpoint.date: 2018-09-07 14:50
              max.dewpoint.val: 14.9
               max.hum_in.date: 2019-08-31 07:04
                max.hum_in.val: 73
              max.hum_out.date: 2007-01-05 18:55
...
...
```

Running this: 
`root@weewx:/home/weewx/bin# ./wee_device /data/weewx.conf --current`
Will give me error messages similar or equal to above but actually readings from the station!
```
{'usUnits': 16, 'dateTime': 1590603593, 'inHumidity': 41.0, 'inTemp': 20.8, 'outHumidity': None, 'outTemp': 13.600000000000001, 'pressure': 1021.4000000000001, 'windSpeed': 0.0, 'windGust': 0.0, 'windDir': 315.0, 'rain': None, 'radiation': None, 'UV': None, 'status': 0, 'ptr': 384, 'delay': 2, 'rxCheckPercent': 100, 'outTempBatteryStatus': 0, 'rainTotal': 484.10999999999996}
```

**Comment by felddy** [comment]
Well that is some progress.   It looks like we've got a permission error to take care of: 

```
May 27 21:25:18 weewx[19] CRITICAL weewx.drivers.fousb: Unable to claim USB interface 0: [Errno 13] Access denied (insufficient permissions)
```

We've already got `--privileged` set, so I'd try setting the `WEEWX_UID` and `WEEWX_GID` to a uid/gid that are in the USB group on your machine.    You could try running at as root (`0`) to test this idea.

**Comment by Nornode** [comment]
Thanks for the help again, and sorry for the slow response!
A bit of life happened in between.

I have changed the startup-config to be:
```
docker create \
  -i \
  -t \
  --privileged \
  --name=$name \
  --hostname=$name \
  --net=host \
  --device=/dev/ttyUSB0 \
  --device=/dev/ttyAMA0 \
  -v /storage/nfs.docker/$name:/data \
  --restart unless-stopped \
  -e TIMEZONE=Europe/Helsinki \
  -e WEEWX_UID=0 \
  -e WEEWX_GID=0 \
    felddy/weewx:latest
```
Disclaimer (probably a bit late now):
My OS is: `LibreELEC (official): 9.2.1`
So a quite limited OS, and the only user & group is 'root' UID & GID = 0.
This puts the container into a (never-ending?) loop:
```
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:46 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
Jun  3 08:48:47 syslogd started: BusyBox v1.30.1
```
After this the host gets unresponsive and it takes ages to get through the 'docker stop weewx' command.

**Comment by felddy** [comment]
Do worries on the delay.  I've fixed the loop bug when requesting uid 0.  A new release has been pushed to the hub.

**Comment by Nornode** [comment]
Thank you for your help! 
I wonder why this won't be the new default weewx image when this is done?... 

_I'm no longer sure if this is an error related to the container, or if this is in my configs somewhere?_
The docker logs complain about the database being locked even if it's just created.

```
Jun  4 08:42:00 syslogd started: BusyBox v1.30.1
Jun  4 08:42:01 weewx[10] INFO __main__: Initializing weewx version 4.0.0
Jun  4 08:42:01 weewx[10] INFO __main__: Using Python 3.8.3 (default, May 20 2020, 19:06:13)  [GCC 8.3.0]
Jun  4 08:42:01 weewx[10] INFO __main__: Platform Linux-4.19.106-armv7l-with-glibc2.4
Jun  4 08:42:01 weewx[10] INFO __main__: Locale is 'C.UTF-8'
Jun  4 08:42:02 weewx[10] INFO __main__: Using configuration file /data/weewx.conf
Jun  4 08:42:02 weewx[10] INFO weewx.engine: Loading station type FineOffsetUSB (weewx.drivers.fousb)
Jun  4 08:42:02 weewx[10] INFO weewx.drivers.fousb: driver version is 1.20
Jun  4 08:42:02 weewx[10] INFO weewx.drivers.fousb: polling mode is PERIODIC
Jun  4 08:42:02 weewx[10] INFO weewx.drivers.fousb: polling interval is 60
Jun  4 08:42:02 weewx[10] INFO weewx.drivers.fousb: found station on USB bus= device=
Jun  4 08:42:02 weewx[10] INFO weewx.engine: StdConvert target unit is 0x1
Jun  4 08:48:26 weewx[10] CRITICAL __main__: Caught unrecoverable exception:
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****  database is locked
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****  Traceback (most recent call last):
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "./bin/weewxd", line 148, in main
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      engine = weewx.engine.StdEngine(config_dict)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 75, in __init__
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      self.loadServices(config_dict)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/engine.py", line 138, in loadServices
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      obj = weeutil.weeutil.get_object(svc)(self,config_dict)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/wxservices.py", line 77, in __init__
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      self.calc = WXCalculate(config_dict,
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/wxservices.py", line 122, in __init__
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      self.db_manager = db_binder.get_manager(
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/manager.py", line 530, in get_manager
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      self.manager_cache[data_binding] = open_manager(manager_dict, initialize)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/manager.py", line 678, in open_manager
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      return manager_cls.open_with_create(manager_dict['database_dict'],
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/manager.py", line 163, in open_with_create
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      dbmanager = cls(connection, table_name=table_name, schema=schema)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/manager.py", line 813, in __init__
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      super(DaySummaryManager, self).__init__(connection, table_name, schema)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weewx/manager.py", line 83, in __init__
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      self.sqlkeys = self.connection.columnsOf(self.table_name)
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weedb/sqlite.py", line 170, in columnsOf
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      column_list = [row[1] for row in self.genSchemaOf(table)]
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weedb/sqlite.py", line 170, in <listcomp>
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      column_list = [row[1] for row in self.genSchemaOf(table)]
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****    File "/home/weewx/bin/weedb/sqlite.py", line 159, in genSchemaOf
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****      for row in self.connection.execute("""PRAGMA table_info(%s);""" % table):
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****  sqlite3.OperationalError: database is locked
Jun  4 08:48:26 weewx[10] CRITICAL __main__:     ****  Exiting.
Traceback (most recent call last):
  File "./bin/weewxd", line 261, in <module>
    main()
  File "./bin/weewxd", line 148, in main
    engine = weewx.engine.StdEngine(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 75, in __init__
    self.loadServices(config_dict)
  File "/home/weewx/bin/weewx/engine.py", line 138, in loadServices
    obj = weeutil.weeutil.get_object(svc)(self,config_dict)
  File "/home/weewx/bin/weewx/wxservices.py", line 77, in __init__
    self.calc = WXCalculate(config_dict,
  File "/home/weewx/bin/weewx/wxservices.py", line 122, in __init__
    self.db_manager = db_binder.get_manager(
  File "/home/weewx/bin/weewx/manager.py", line 530, in get_manager
    self.manager_cache[data_binding] = open_manager(manager_dict, initialize)
  File "/home/weewx/bin/weewx/manager.py", line 678, in open_manager
    return manager_cls.open_with_create(manager_dict['database_dict'],
  File "/home/weewx/bin/weewx/manager.py", line 163, in open_with_create
    dbmanager = cls(connection, table_name=table_name, schema=schema)
  File "/home/weewx/bin/weewx/manager.py", line 813, in __init__
    super(DaySummaryManager, self).__init__(connection, table_name, schema)
  File "/home/weewx/bin/weewx/manager.py", line 83, in __init__
    self.sqlkeys = self.connection.columnsOf(self.table_name)
  File "/home/weewx/bin/weedb/sqlite.py", line 170, in columnsOf
    column_list = [row[1] for row in self.genSchemaOf(table)]
  File "/home/weewx/bin/weedb/sqlite.py", line 170, in <listcomp>
    column_list = [row[1] for row in self.genSchemaOf(table)]
  File "/home/weewx/bin/weedb/sqlite.py", line 159, in genSchemaOf
    for row in self.connection.execute("""PRAGMA table_info(%s);""" % table):
**sqlite3.OperationalError: database is locked**
```

relevant(?) part of my weewx.conf (in /data)
```
[DatabaseTypes]
    
    # Defaults for SQLite databases
    [[SQLite]]
        driver = weedb.sqlite
        SQLITE_ROOT = /data
```

The DB file is created but maintains a size of zero:
`-rw-r--r--    1 root     root           0 Jun  3 23:00 weewx.sdb`

**Comment by Nornode** [comment]
After further research I can't find anything that points to this not being something in the container itself. @felddy could you think if anything that might prevent writes/db locks?
